src/event-log/event-log-level.ts

```diff
@@ -15,101 +18,114 @@ type EventLogLevelConfig = {
-    this.config = {
+    this.index = new IndexLevel({
       location: 'EVENTLOG',
       createLevelDatabase,
       ...config,
-    };
-
-    this.db = new LevelWrapper<string>({
-      location            : this.config.location!,
-      createLevelDatabase : this.config.createLevelDatabase,
-      valueEncoding       : 'utf8',
     });
+
     this.ulidFactory = monotonicFactory();
   }
 
   async open(): Promise<void> {
-    return this.db.open();
+    return this.index.open();
   }
 
   async close(): Promise<void> {
-    return this.db.close();
+    return this.index.close();
   }
 
   async clear(): Promise<void> {
-    return this.db.clear();
+    return this.index.clear();
   }
 
-  async append(tenant: string, messageCid: string): Promise<string> {
-    const tenantEventLog = await this.db.partition(tenant);
-    const watermarkLog = await tenantEventLog.partition(WATERMARKS_SUBLEVEL_NAME);
-    const cidLog = await tenantEventLog.partition(CIDS_SUBLEVEL_NAME);
-
+  /**
+   * Appends Events to the EventLog in the order they are appended using a ulid watermark.
+   * The indexes are used to query the event-log for specific events.
+   *
+   * @param tenant
+   * @param messageCid the messageCid of the message that is being appended to the log.
+   * @param indexes property and value indexes to use for querying potential events.
+   * @returns the ulid watermark generated during this operation.
+   */
+  async append(tenant: string, messageCid: string, indexes: { [key:string]: unknown }): Promise<string> {
     const watermark = this.ulidFactory();
-
-    await watermarkLog.put(watermark, messageCid);
-    await cidLog.put(messageCid, watermark);
-
+    // A reverse lookup in order to delete by messageCid
+    // We are using the watermark as the key for this index.
+    // When deleting by messageCid we need to look up the watermark.
+    const cidLogIndex = await this.messageCidPartition(tenant);
+    await cidLogIndex.put(messageCid, watermark);
+
+    const event:Event = { messageCid, watermark };
+    await this.index.put(tenant, watermark, event, { ...indexes, watermark });
     return watermark;
   }
 
-  async getEvents(tenant: string, options?: GetEventsOptions): Promise<Event[]> {
-    const tenantEventLog = await this.db.partition(tenant);
-    const watermarkLog = await tenantEventLog.partition(WATERMARKS_SUBLEVEL_NAME);
-    const events: Array<Event> = [];
-
-    for await (const [key, value] of watermarkLog.iterator(options)) {
-      const event = { watermark: key, messageCid: value };
-      events.push(event);
-    }
+  /**
+   * Queries Events for a given tenant using the filters provided.
+   *
+   * @param tenant
+   * @param filters an array of filters that designates which event properties are being queried.
+   * @returns an array of matching Events without duplicate entries between the filters.
+   */
+  async queryEvents(tenant: string, filters: Filter[], watermark?: string): Promise<Event[]> {
```

Comment:
Sanity: no paging? Same question to `getEvents()`

Reply:
Will add this in a separate PR along with ProtocolQuery pagination (that would be all the queries).